Implement a code formatter for GDScript in the script editor

Razoric480 opened the pull request:
This implements godotengine/godot-proposals#3630 and puts in a code formatter in the GDScript module. This is accessible via Edit/Format Code, Alt + Shift + F, or when saving while having the new _Format On Save_ editor setting enabled.

The formatting is based off of the [official style guide](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_styleguide.html), though there are cases where automation may cause differences. Or coding issues/bugs to be fixed in the formatter :3

Comments are supported. This involved making the GDScriptTokenizer recognize and return comment tokens instead of bypassing them. The GDScriptParser applies those comments in the various parse\_ nodes as header comments (above a line of code), inline comment (at the end of a line of code), or footer comment (at the bottom of a suite or class). They are then printed by the formatter as it goes through the code.

Adds three new editor settings:
- `text_editor/behavior/formatter/format_on_save` - defaults to `false`. When saving scripts, whether to automatically run the formatter.
- `text_editor/behavior/formatter/indent_in_multiline_block` - defaults to `2`. When wrapping a block of code or parameters inside of parenthesis, how many indents to add from the first.
- `text_editor/behavior/formatter/lines_between_functions` - defaults to `2`. It is the number of blank lines between functions.

- `text_editor/appearance/guidelines/line_length_guideline_hard_column` is used for the target line length, which defaults to `100`.

There's a suite of unit tests for formatting issues. It could maybe be moved into a test runner later on to make the code cleaner, and possibly some tests merged and consolidated, though I used tests to do a fair bit of TDD.

### Likely improvements required

- I can't forsee all scripts, so odds are there will be various formatting issues that come from edge cases or code combinations that happen to cause bigger than normal gaps, disappearing comments that I forgot to account for, or various other issues like that.
- Extra features, like black's trailing commas forcing wraps, could be in subsequent PRs

### Pushed to subsequent PRs

- Integration with CLI
  - Since the formatter lives in a module, it can't/shouldn't really be called or instanced in main.cpp where all the CLI stuff happens. It could happen in gdscript.cpp's init(), but the question then is how to make the editor behave like the other CLI options, where the editor then shuts itself down after it finishes the task.
  - Other issue: passing options to the formatter, like indent type. I guess if it's running in tool mode it can just get the editor settings, but could potentially be called to be overwritten.
- Integration with GDScript language server
  - https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_formatting

Razoric480 committed:
Add script formatter, comment parsing in parser, and editor settings

* Add verbose messages for debugging purposes
* Add disabled lines handling to parser
* Force newline at end of code

Scony commented:
Looks really good! @Razoric480 Are you considering harnessing `GDScript` testing framework to it?

Razoric480 commented:
I didn't make immediate plans for it, but I can certainly investigate adding some tests.

gustavi commented:
First, thank you **a lot** for working on this feature. As a Python developer `black` make my life (as a team lead) easier and resolved a lot of "conflicts" on *the good way* of doing thinks. I like the idea of "it's not perfect but it's the same for everyone".

As someone using `black` since the first version my feedback will compare the result with this tool. Since Python and GDScript have almost the same type of syntax I thing this is relevant.

I know you are working on some features (like comments) but I'll put my feedback here anyway, you can use my samples for testing if you want!

# Feedback

## Some line breaks around comments are removed

```diff

-##
-## XXX
-##
-
-
+# #
+# # XXX
+# #
 const VERSION := "xxx"

-
 # ----- Debug ------------------------------------------------------------------
-
-
```

I guess it's a bug.

## [Documentation comments](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_documentation_comments.html) are not supported

```diff
-##
-## XXX
-##
-
-
+# #
+# # XXX
+# #
```

## Extra parenthesis are added around constant declaration

Example 1:

```diff
 const VERSION := "XXX"

-
 # ----- Debug ------------------------------------------------------------------
-
-
-const DEBUG := false
+const DEBUG := (
+    false
+)
 const DEBUG_IS_VERBOSE := false
 const DEBUG_SERVER := DEBUG
 const DEBUG_CLIENT := DEBUG
```

Example 2:

```diff
-
 # ----- Generic ----------------------------------------------------------------
-
-
-const DEFAULT_ANIMATIONS_FPS := 30
+const DEFAULT_ANIMATIONS_FPS := (
+    30
+)
 const FPS_XXX := 20.0
```

## Comments at the end of a file are removed

## Preferred char for indenting is ignored

I use 4 spaces for indent (Editor setting) but the reformat force tabs. I don't know if it's a bug or not.

## 3 lines are used between functions instead of 2

cf https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_styleguide.html#blank-lines

## Single quotes strings are replaced by double quotes, even when double quotes are present inside the String

```diff
 func _attrib_as_string() -> String:
     if attrib:
-        var rv: = ""
+        var rv := ""
         for key in keys():
-            rv += ' {key}="{value}"'.format(
-                {"key": key, "value": get_attr(key)}
-            )
+            rv += " {key}="{value}"".format({"key": key, "value": get_attr(key)})
         return rv
     return ""
```

# Crash report

I'm able to crash the editor (Windows 10 - v4.0.alpha.custom_build [1edd7291e]) with the following scripts:

```gdscript
extends RefCounted
# bbb
var root_element: AStar2D
# aaa
```

```gdscript
class_name FooBar

func _init(tag_param: String):
    assert(tag_param != "")
```

```
class_name XMLElement2
extends RefCounted


func _as_string(indent: int) -> String:
    var element_indent: String
    if indent == -1:
        indent = -2  # small hack for for subelements
        element_indent = ""
    else:
        element_indent = "  ".repeat(indent)

    return ""
```

# Features suggestion

- Add a setting "Reformat on save"

That's all for the moment! I'll test more on next update but I had a lot of crashes (95% of my files). Thank you again for your work.

Razoric480 commented:
- Docstring still need to be implemented, just working on in-line for now.
- Such short `const` being broken is weird O_o I'll have a look.
- Comments at end of file removed... interesting - guess they don't have any object to attach to and get discarded.
- Right now tabs are hard coded, but changing spacing and style will go into Editor Settings integration.
- I'd forgotten about single quotes. Will make sure to add support for them.
- Crashes, haven't run into any yet, but I've mostly been running tests. Thanks for the example script - I'll use it to make sure things don't explode as I go.
- Reformat on saves is planned :)

Thanks for the review~

Razoric480 commented:
Progress update: This is a little blocked due to lack of time - near full time job makes it a bit tricky to raise the energy to tackle the project. If it wasn't for comments this would be a lot more done, but comments are a very complex endeavour given the parser isn't particularly helpful with them.

vnen commented:
We checked this briefly during the GDScript meeting. The consensus is that this idea is good (along with the proposal), we just need to flesh out this PR to a mergeable state. It doesn't need to have all features at first but it needs to have a base so we can work on it, fixing some issues like the problems with comments for instance. I can help, just ping me on the Godot Rocket chat and we can talk.

On a personal note, sorry it took me so long to check this out. Thanks for working on it.

vnen commented:
BTW, I haven't really took a look in the implementation itself, but if something is blocking you we may discuss some other approach, either in your code or in what's there already in GDScript. If something is missing from the GDScript side we can add it.

Razoric480 commented:
Big update pushes the PR into a ready state. It'll likely have some formatting oddities and/or bugs, cause you can't account for every script or combination of lines of code, but it gets it the majority of the way.

derammo commented on modules/gdscript/gdscript_format.h:
```diff
@@ -0,0 +1,201 @@
+	static bool node_has_comments(const GDP::Node *p_node);
+	static bool children_have_comments(const GDP::Node *p_parent);
+
+	static int get_operation_priority(const GDP::BinaryOpNode::OpType p_op_type) {
```
look at `GDScriptParser::get_rule(token)`

I believe a BinaryOpNode knows what precedence value is via `get_rule(current.type)->precedence`

That's just based on looking in the code just now, so it might not work out.  But it is something you probably need to explain why the formatter has its own opinion about the language definition, if you can't make it work with the data that is in the tree you are working with.

Maybe it works though then you could dramatically simplify.

Razoric480 replied:
get_rule() operates off of GDScriptTokenizer's Token's Type enum, which does not match BinaryOpNode's Type list, and there is no mechanic to reverse engineer the token from which a Node came from.

So ultimately, using get_rule would:
- need to store the parser in a class variable
- need a function to translate the type of binary operator optype into a token type
- get the rule from the parser using the translation

So I feel it'd more complicated to set up the architecture there than it is just having a precedence rule function for this one node.

derammo replied:
I agree.  I had searched poorly.  The code was referring to the token 'current' that is only transient during parsing, and the node never stores the precedence that it dereferences there.   That's too bad.   It still feels to me the facility you created is weird to exist outside the language definition.  If it is in fact possible to infer precedence JUST From a node type, then it should be possible to just have a table that maps from node types to token types and then look up the token in the parser.  That way, if the language changes you would get updates.   If you mapped multiple tokens, you could even check if there is any contradiction in your code.  Like if the language actually does NOT strictly sort those BinaryOpNode types because of context.  Just some ideas you might or might not find interesting.

aaronfranke commented on editor/editor_settings.cpp:
```diff
@@ -551,6 +551,12 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	_initial_set("text_editor/behavior/indent/auto_indent", true);
 
+	// Behavior: Formatter
+	_initial_set("text_editor/behavior/formatter/maximum_line_length", 80);
```
We should change the default to 100. It's sensible to suggest users to keep it 80 or less, but this doesn't make sense to enforce automatically. Game logic in GDScript will naturally have some long lines.

The script editor already shows a faint line at 80 characters and a thicker line at 100 characters. I think the character count used by the formatter should be the same as this thicker line. Actually, what we should likely do is unify these settings. `text_editor/appearance/guidelines/line_length_guideline_hard_column`

aaronfranke commented:
I went to go test this on my company's large project, but it crashes.

```
ERROR: Parser bug: Mismatch in extents tracking stack.
   at: complete_extents (modules/gdscript/gdscript_parser.cpp:4344)
```

<details>

```
ERROR: Parser bug: Mismatch in extents tracking stack.
   at: complete_extents (modules/gdscript/gdscript_parser.cpp:4344)

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.0.alpha.custom_build (6675c25fac48209609b7c11b2048279c61b1b80b)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] 1   libsystem_platform.dylib            0x00000001aecd74a4 _sigtramp + 56
[2] GDScriptParser::parse_match() (in godot.macos.opt.tools.arm64) (gdscript_parser.cpp:2122)
[3] GDScriptParser::parse_match() (in godot.macos.opt.tools.arm64) (gdscript_parser.cpp:2122)
[4] GDScriptParser::parse_statement() (in godot.macos.opt.tools.arm64) (gdscript_parser.cpp:0)
[5] GDScriptParser::parse_suite(String const&, GDScriptParser::SuiteNode*, bool) (in godot.macos.opt.tools.arm64) (gdscript_parser.cpp:0)
[6] GDScriptParser::parse_function() (in godot.macos.opt.tools.arm64) (gdscript_parser.cpp:1561)
[7] void GDScriptParser::parse_class_member<GDScriptParser::FunctionNode>(GDScriptParser::FunctionNode* (GDScriptParser::*)(), GDScriptParser::AnnotationInfo::TargetKind, String const&) (in godot.macos.opt.tools.arm64) (gdscript_parser.cpp:0)
[8] GDScriptParser::parse_class_body(bool) (in godot.macos.opt.tools.arm64) (gdscript_parser.cpp:909)
[9] GDScriptParser::parse_program() (in godot.macos.opt.tools.arm64) (gdscript_parser.cpp:674)
[10] GDScriptParser::parse(String const&, String const&, bool) (in godot.macos.opt.tools.arm64) (gdscript_parser.cpp:402)
[11] GDScriptParserRef::raise_status(GDScriptParserRef::Status) (in godot.macos.opt.tools.arm64) (gdscript_cache.cpp:62)
[12] GDScriptAnalyzer::resolve_datatype(GDScriptParser::TypeNode*) (in godot.macos.opt.tools.arm64) (gdscript_analyzer.cpp:488)
[13] GDScriptAnalyzer::resolve_class_interface(GDScriptParser::ClassNode*) (in godot.macos.opt.tools.arm64) (gdscript_analyzer.cpp:629)
[14] GDScriptAnalyzer::resolve_interface() (in godot.macos.opt.tools.arm64) (gdscript_analyzer.cpp:4240)
[15] GDScriptParserRef::raise_status(GDScriptParserRef::Status) (in godot.macos.opt.tools.arm64) (gdscript_cache.cpp:75)
[16] GDScriptAnalyzer::resolve_datatype(GDScriptParser::TypeNode*) (in godot.macos.opt.tools.arm64) (gdscript_analyzer.cpp:488)
[17] GDScriptAnalyzer::resolve_function_signature(GDScriptParser::FunctionNode*) (in godot.macos.opt.tools.arm64) (gdscript_analyzer.cpp:1167)
[18] GDScriptAnalyzer::resolve_class_interface(GDScriptParser::ClassNode*) (in godot.macos.opt.tools.arm64) (gdscript_analyzer.cpp:828)
[19] GDScriptAnalyzer::analyze() (in godot.macos.opt.tools.arm64) (gdscript_analyzer.cpp:4267)
[20] GDScript::reload(bool) (in godot.macos.opt.tools.arm64) (gdscript.cpp:884)
[21] GDScriptCache::get_full_script(String const&, Error&, String const&) (in godot.macos.opt.tools.arm64) (gdscript_cache.cpp:209)
[22] ResourceFormatLoaderGDScript::load(String const&, String const&, Error*, bool, float*, ResourceFormatLoader::CacheMode) (in godot.macos.opt.tools.arm64) (gdscript.cpp:2336)
[23] ResourceLoader::_load(String const&, String const&, String const&, ResourceFormatLoader::CacheMode, Error*, bool, float*) (in godot.macos.opt.tools.arm64) (resource_loader.cpp:214)
[24] ResourceLoader::_thread_load_function(void*) (in godot.macos.opt.tools.arm64) (resource_loader.cpp:240)
[25] ResourceLoader::load(String const&, String const&, ResourceFormatLoader::CacheMode, Error*) (in godot.macos.opt.tools.arm64) (resource_loader.cpp:570)
[26] ResourceLoaderText::load() (in godot.macos.opt.tools.arm64) (resource_format_text.cpp:482)
[27] ResourceFormatLoaderText::load(String const&, String const&, Error*, bool, float*, ResourceFormatLoader::CacheMode) (in godot.macos.opt.tools.arm64) (resource_format_text.cpp:1568)
[28] ResourceLoader::_load(String const&, String const&, String const&, ResourceFormatLoader::CacheMode, Error*, bool, float*) (in godot.macos.opt.tools.arm64) (resource_loader.cpp:214)
[29] ResourceLoader::_thread_load_function(void*) (in godot.macos.opt.tools.arm64) (resource_loader.cpp:240)
[30] ResourceLoader::load(String const&, String const&, ResourceFormatLoader::CacheMode, Error*) (in godot.macos.opt.tools.arm64) (resource_loader.cpp:570)
[31] EditorAutoloadSettings::_create_autoload(String const&) (in godot.macos.opt.tools.arm64) (editor_autoload_settings.cpp:402)
[32] EditorAutoloadSettings::EditorAutoloadSettings() (in godot.macos.opt.tools.arm64) (editor_autoload_settings.cpp:859)
[33] ProjectSettingsEditor::ProjectSettingsEditor(EditorData*) (in godot.macos.opt.tools.arm64) (project_settings_editor.cpp:667)
[34] EditorNode::EditorNode() (in godot.macos.opt.tools.arm64) (editor_node.cpp:6469)
[35] Main::start() (in godot.macos.opt.tools.arm64) (main.cpp:2508)
[36] main (in godot.macos.opt.tools.arm64) (godot_main_macos.mm:85)
[37] 37  dyld                                0x000000010c1f108c start + 520
-- END OF BACKTRACE --
```

</details>

Razoric480 commented:
> I went to go test this on my company's large project, but it crashes. @aaronfranke 

Hrm. Without a script that makes it tricky to debug. Looks like it's happening while parsing `match:`.

aaronfranke commented:
@Razoric480 Perhaps the formatter should print out what script it's on when it crashes? Otherwise it's not trivial for me to find out which script is the problem when I have a project with several hundred scripts in it.

Razoric480 commented:
@aaronfranke I've added a temporary call in GDScriptParser to check for the editor running with the `--verbose` flag. If it's crashing over a file, it should output the file. If it's crashing over raw code (from the formatter itself) it should output the source code.

Mind that it will output everything running through the parser, so it'll be spammy, but the one right before the crash is the one you'd be looking for.

aaronfranke commented:
The crash can be reproduced with the following code (after typing `0`, it crashes)

```gdscript
extends Node

func _ready() -> void:
	var x = 0
	match x:
		0
```

I discovered a non-crash bug with code disabled with <kbd>Ctrl</kbd>+<kbd>K</kbd>:

![script-comment-bug](https://user-images.githubusercontent.com/1646875/182247902-8d42d765-2a67-44dc-81df-a71ccf16d914.png)

Razoric480 commented:
Thanks! Will look into it.

Razoric480 commented:
@aaronfranke I should have fixed the crash on malformed match, and the disabled-line errors, so it should be good for a second go around.

Razoric480 commented:
@aaronfranke Boy there sure is a lot of ways to use comments! It's almost as if it's a non-trivial problem :P Should be fixed, and I've prematurely fixed it for dictionaries as well.

aaronfranke commented:
@Razoric480 Indeed! I can confirm that the latest update to this PR is able to open my company's project without crashing. However, it still does not run, here's another test case:

<img width="600" alt="Screen Shot 2022-08-02 at 10 13 16 PM" src="https://user-images.githubusercontent.com/1646875/182516259-77fa5019-6a4f-45bd-ace3-bd362073a8b2.png">

Razoric480 commented:
I'm alive again! With a new update! @aaronfranke If you want to give it another whirl, last issue should now be fixed ~~and I'm ready for the next bug you'll inevitably find :P~~

aaronfranke commented:
NodePath literals are destroyed. Saving this causes the `^` and `"` to be stripped:

```gdscript
@onready var child = get_node(^"Child")
```

This code causes a parser error:

```gdscript
	var arr = ["string"
		# Comment
	]
```

Array formatting inside of `append_array` is really weird. It becomes this:

```gdscript
func _ready():
	var arr = []
	arr.append_array(
			[
						"long string 1", "long string 2", "long string 3", "long string 4", "long string 5", "long string 6"
					]
	)
```

---

We should probably disallow annotations being on the same lines as function declarations. In this PR, this:

```gdscript
@rpc(call_remote, any_peer, reliable)
func my_rpc(parameter_one, parameter_two, parameter_three, parameter_four):
	pass
```

Becomes this:

```gdscript
@rpc(call_remote, any_peer, reliable) func my_rpc(parameter_one, parameter_two, parameter_three, parameter_four):
	pass
```

---

The formatter is really eager to move things onto one line, which seems sub-optimal. It automatically turns this:

```gdscript
enum MyEnum {
	MEMBER_ONE = 1,
	MEMBER_TWO = 2,
	MEMBER_THREE = 3,
	MEMBER_FOUR = 4,
	MEMBER_FIVE = 5,
}
```

Into this:

```gdscript
enum MyEnum { MEMBER_ONE = 1, MEMBER_TWO = 2, MEMBER_THREE = 3, MEMBER_FOUR = 4, MEMBER_FIVE = 5 }
```

Razoric480 commented:
> NodePath literals are destroyed. Saving this causes the ^ and " to be stripped:
Oh, strange. I'll add some test cases and fix it up.

> Array formatting inside of append_array is really weird. It becomes this:
=o Very weird. Will look into it.

> We should probably disallow annotations being on the same lines as function declarations.
I've thought about that. Originally I made some rules about how I'd decide, but it adds complexity. Probably easier to disallow entirely.

> The formatter is really eager to move things onto one line, which seems sub-optimal. It automatically turns this:
That's unfortunately a bit tricky. The parser doesn't really retain any information about the code - just that it's an enum with a name, and it has 5 elements that have explicit values. By the time the formatter gets to it, it just knows that it's less than 100 lines. It doesn't know if it was written pre-split on multiple lines, or written entirely on one line. This one happens to be 98 characters.

One feature that has been pushed to subsequent PR is to have the parser record the use of trailing commas, which would then force splits in the formatter.

For the rest, thanks for the test cases - will investigate them.

YuriSizov commented:
For the record, I think we'll be okay to merge this once it's stable and had been reviewed by @Paulb23 for the editor integration part. Overall design seems fine, and the feature itself is tangential enough that it wouldn't be a huge problem if it's misbehaving in parts at the merge time.

@Razoric480 Consider adding the test cases from Aaron to the unit test suite.

Razoric480 commented:
New update, fixing NodePath and StringName literals, parser error on single element with footer, annotations for functions auto-wrap, and call arguments adding too many tabs. Couldn't repro the nested if extra lines, but I added a test for it so might be I just fixed it while doing other things.

Fingers crossed that's the most of it!

aaronfranke commented:
> Couldn't repro the nested if extra lines

I can still reproduce this. There are also still a lot more errors.

Constants are destroyed by the formatter. When I format this code:

```gdscript
var f = fposmod(5, TAU)
```

It automatically changes to this:


```gdscript
var f = fposmod(5, 6.28318530717959)
```

This code gives an error of unexpected indent:

```gdscript
func _ready():
	if true:
		pass
# comment
	pass
```

The formatter destroys the type information for typed arrays (it removes the [Node] part):

```gdscript
@onready var children: Array[Node] = get_children()
```

Comment indentation is not preserved in all cases. The formatter adds an extra tab before this comment:

```gdscript
func _ready():
	if true:
		pass
	# This comment is for the print statement.
	print("hi")
```

Signals that explicitly have zero arguments are stripped of their parenthesis (possibly intentional, but we have been using `()` in our codebase to explicitly mark these as such, so I'd argue against removing them):

```gdscript
signal my_signal()
```

Any comments above `@onready` vars are deleted:

```gdscript
## I am a comment describing var hi
@onready var hi
```

---

The formatter currently formats array arguments like this:

```gdscript
func _ready():
	var arr = []
	arr.append_array(
			[
				"test with a long string 1",
				"test with a long string 2",
				"test with a long string 3",
				"test with a long string 4"
			]
	)
```

The formatter should keep trailing commas on multi-line lists. This is what the style guide says [here](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_styleguide.html#trailing-comma).

Also, I would prefer to avoid having opening and closing brackets by themselves:

```gdscript
func _ready():
	var arr = []
	arr.append_array([
		"test with a long string 1",
		"test with a long string 2",
		"test with a long string 3",
		"test with a long string 4",
	])
```

Razoric480 commented:
Thanks, as always! Will start plugging these into tests and fixing.

Scony commented:
@Razoric480 is there an easy way to run formatter vs multiple files?
I'm asking because I got quite a lot of `GDScript` files here: https://github.com/Scony/godot-gdscript-toolkit/tree/4.0-int/tests/formatter/input-output-pairs (not to mention entire repo) but testing everything by hand is cumbersome.

Razoric480 commented:
> @Razoric480 is there an easy way to run formatter vs multiple files? I'm asking because I got quite a lot of `GDScript` files here: https://github.com/Scony/godot-gdscript-toolkit/tree/4.0-int/tests/formatter/input-output-pairs (not to mention entire repo) but testing everything by hand is cumbersome.

@Scony 

At the moment, no - CLI access is planned for a future PR, as this one was running long in the teeth and I wanted to at least get the start of it in the early beta days. So this is the "save on format" and "code editor" PR. Then I will do CLI as a matter of priority next in a follow up PR, and finally language server intergration.

Once CLI is in, a single `**/*.gd` blob should be able to cover everything.

Paulb23 commented on modules/gdscript/gdscript_format.cpp:
```diff
@@ -0,0 +1,1958 @@
+
+	GDP::ClassNode *root = parser.get_tree();
+
+	String output = "";
```
With all this string manipulation, would it be worth looking at `StringBuilder`?

Paulb23 reviewed (commented):
Can't comment too much on the `GDScript` side of things.

Missing a format on save under:
```c++
void ScriptEditor::_menu_option(int p_option) {
	FILE_SAVE_AS
}
```

I'm also slightly concerned about performance here on larger scripts as `set_text` will cause everything to recalculate.

I don't think it's a good idea to link it to `TextEdit` like with convert indent and trim trailing whitespace. However, if `format_on_save` is true, can we skip those calls so we don't end up formatting the same script three times?

Secondly, could we take a page out of `EDIT_AUTO_INDENT` book, and allow formatting parts of the script, or perhaps only call `set_line` on parts of the script that have changed?

Razoric480 commented:
Big ol' update.
- Went through @aaronfranke 's latest findings and fixed those.
- Replaced String with StringBuilder  as suggested by PaulB
- Wrapped the other on-save actions inside of a not-format-on-save block

Ready for the next set of issues :P

I'm a bit unsure how to go about fixing the caret situation, however. I guess we could analyze the number of lines that came before the caret and reduce by the difference if it's different and maybe do the same with columns. I'll have to give it a think.

aaronfranke commented:
@Razoric480 Nothing happens when I save a script with the current version of this PR. Format on save is enabled in editor settings. I tested both this PR as-is, and rebased on the latest master, both do not do any formatting.

Razoric480 commented:
Oh, uhm. Oops. I'll make sure to fix that.

aaronfranke commented:
This code fails to parse (Expected statement, found "else" instead):

```gdscript
func _ready():
	if true:
		pass
# Comment
	else:
		pass
```

This code fails to parse (Unindent doesn't match the previous indentation level):

```gdscript
func _ready():
	if true:
		pass
# Comment
		pass
	pass
```

The bug with a lot of extra newlines when leaving a scope is still present:

```gdscript
func _ready():
	if true:
		if true:
			if true:
				if true:
					pass




	pass
```

---

```gdscript
func _ready():
	var this_is_a_very_long_boolean_for_test_purposes: bool = false
	if this_is_a_very_long_boolean_for_test_purposes or this_is_a_very_long_boolean_for_test_purposes or this_is_a_very_long_boolean_for_test_purposes:
		pass
```

The formatter does correctly split the above into multiple lines, but it adds multiple sets of parenthesis for no reason:

```gdscript
func _ready():
	var this_is_a_very_long_boolean_for_test_purposes: bool = false
	if (
		(
			this_is_a_very_long_boolean_for_test_purposes or this_is_a_very_long_boolean_for_test_purposes
			or this_is_a_very_long_boolean_for_test_purposes
		)
	):
		pass
```

---

```gdscript
var my_array = [
	"there is a bug with an extra newline at the end of arrays but only when the contents have long lines",
	# Comment

]
```

Razoric480 commented:
New update:
- Replaced the use of `set_text` with `set_line`
- Moved disabled line handling into the parser, so that should reduce the number of errors with those
- Fixed formatting issues found by @aaronfranke 's very gracious and patient testing

I did manage to recreate the scope extra line bug so that should hopefully be fixed properly now. Fewer parentheses around binary operators, too.

aaronfranke commented:
Sorry for the delay, here's the next round of problems found by testing my company's codebase.

I can still reproduce the extra lines after the end of blocks issue:

```gdscript
func _ready():
	for i in range(5):
		for j in range(5):
			for k in range(5):
				pass




	pass


# Comment
```

The formatter is removing the space before this equal sign:

```gdscript
func my_func(param = true):
	pass
```

The formatter is expanding `\n`. This code:

```gdscript
func _ready():
	print("hi\n")
```

Becomes this:

```gdscript
func _ready():
	print("hi
")
```

---

Property setters defined using methods are broken after this PR formats them. This:

```gdscript
var my_str: String = "":
	set = set_my_str


func set_my_str(value: String):
	pass
```

Becomes this (complains that it didn't expect an identifier of setget):

```gdscript
var my_str: String = "" setget set_my_str


func set_my_str(value: String):
	pass
```

---

The formatter breaks this code, where addition is performed and then abs is called on the result:

```gdscript
func _ready():
	var a = Vector2(1, 2)
	var b = Vector2(3, 4)
	var c = (a + b).abs()
```

It becomes this code, where abs is called on b and then it's added to a:

```gdscript
func _ready():
	var a = Vector2(1, 2)
	var b = Vector2(3, 4)
	var c = a + b.abs()
```

---

The formatter is duplicating a comment at the end of a line in this case:

```gdscript
var grandchild = $Child.get_node(^"Grandchild") # Comment
```

After formatting a few times, it looks like this:

```gdscript
var grandchild = $Child.get_node(
		^"Grandchild"
) # Comment # Comment # Comment # Comment # Comment # Comment # Comment # Comment # Comment # Comment # Comment # Comment # Comment # Comment # Comment # Comment
```

---

This code fails to parse. And yes it's isn't a well-formatted snippet of code to start with, but it should at least parse.

```gdscript
func _ready():
	var x: int = 0
	match x:
		0:
			pass
	# comment
			pass
		1:
			pass
```

The formatter is indenting this comment, seems un-ideal (shown here is the before):

```gdscript

func _ready():
	if true:
		pass
	# This comment is describing the below if.
	if true:
		pass
```

The formatter is interpreting this comment as a doc comment, moving it down and adding a space (so the line starts `#`, space, tab). My suggestion is that lines starting with `#` and tab are treated as part of the previous scope.

```gdscript
func _ready():
	pass
#	pass


func other_func():
	pass
```

Can you add a feature to turn off automatic formatting selectively, for a given line until the end of the file or until it's enabled again? Just like how clang-format has the `/* clang-format off */` comment. Something like:

```gdscript
@formatter(false)
func plane_yz() -> Plane: return Plane(transform.basis.x.normalized(), transform.origin)
func plane_xz() -> Plane: return Plane(transform.basis.y.normalized(), transform.origin)
func plane_xy() -> Plane: return Plane(transform.basis.z.normalized(), transform.origin)
@formatter(true)
```

The syntax can be anything, enable_formatter, skip_format, etc, I don't really have a strong opinion. Also, this is a low-priority, if this is merged without a way to selectively disable formatting then I'll be fine with that.

NathanLovato commented:
I just started testing this, I can confirm the extra lines after block. This:

```gd
func _ready() -> void:
	if not player:
		set_physics_process(false)
		return
	initial_offset = player.global_position - global_position
```

Becomes that (2 inserted lines):

```gd
func _ready() -> void:
	if not player:
		set_physics_process(false)
		return


	initial_offset = player.global_position - global_position
```

NathanLovato commented:
A parenthesis gets put on a newline depending on the following function. This formats correctly:

```gdscript
func _ready() -> void:
	target_detector_3d.area_entered.connect(func(area: Area3D):
		_target = area)
	target_detector_3d.area_exited.connect(func(area: Area3D):
		_target = null)


func _physics_process(delta: float) -> void:
#	if _target:
	pivot.look_at(_target.global_position)
	if timer.is_stopped():
		shoot()
		timer.start()
```

However, if I uncomment the `if _target:` line and indent the following lines, this happens (notice the last parenthesis in the `_ready()` function):

```gdscript
func _ready() -> void:
	target_detector_3d.area_entered.connect(func(area: Area3D):
		_target = area)
	target_detector_3d.area_exited.connect(func(area: Area3D):
		_target = null
)


func _physics_process(delta: float) -> void:
	if _target:
		pivot.look_at(_target.global_position)
		if timer.is_stopped():
			shoot()
			timer.start()
```

dalexeev commented:
Superseded by #76211.
